src/analysis/processing/qgsalgorithmrastersampling.cpp

```diff
@@ -0,0 +1,166 @@
+
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "COLUMN_PREFIX" ), QObject::tr( "Output column prefix" ), QStringLiteral( "SAMPLE_" ), false, true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Sampled" ), QgsProcessing::TypeVectorPoint ) );
+}
+
+bool QgsRasterSamplingAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT_RASTER" ), context );
+  if ( !layer )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "INPUT_RASTER" ) ) );
+
+  mBandCount = layer->bandCount();
+  mCrs = layer->crs();
+  mDataProvider.reset( static_cast< QgsRasterDataProvider *>( layer->dataProvider()->clone() ) );
+
+  return true;
+}
+
+QVariantMap QgsRasterSamplingAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  QString fieldPrefix = parameterAsString( parameters, QStringLiteral( "COLUMN_PREFIX" ), context );
+  QgsFields newFields;
+  for ( int band = 1; band <= mBandCount; band++ )
+  {
+    newFields.append( QgsField( QStringLiteral( "%1%2" ).arg( fieldPrefix, QString::number( band ) ), QVariant::Double ) );
+  }
+  QgsFields fields = QgsProcessingUtils::combineFields( source->fields(), newFields );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields,
+                                          source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  const long count = source->featureCount();
+  const double step = count > 0 ? 100.0 / count : 1;
+  long current = 0;
+
+  QgsCoordinateTransform ct( source->sourceCrs(), mCrs, context.transformContext() );
+  QgsFeatureIterator it = source->getFeatures( QgsFeatureRequest() );
+  QgsFeature feature;
+  while ( it.nextFeature( feature ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( current * step );
+
+    if ( !feature.hasGeometry() )
+    {
+      current++;
+      continue;
+    }
+
+    QgsGeometry geometry = feature.geometry();
+    if ( geometry.isMultipart() && geometry.get()->partCount() > 1 )
+    {
+      sink->addFeature( feature, QgsFeatureSink::FastInsert );
+      feedback->reportError( QObject::tr( "Impossible to sample data of multipart feature %1." ).arg( feature.id() ) );
+      continue;
+    }
+    QgsPointXY point = geometry.isMultipart() ? geometry.asMultiPoint()[0] : geometry.asPoint();
```

Comment:
```suggestion
    QgsPointXY point( geometry.isMultipart() ? qgs_geometrycast< const QgsPoint* >( qgs_geometrycast< const QgsMultiPoint* >( geometry.constGet() )->geometryN( 0 ) ) : qgs_geometrycast< const QgsPoint* >( geometry.constGet() ) );
```

Just a bit more efficient!

TODO: we really need to improve the api for the QgsGeometryCollection subclasses. It's not at all friendly that retrieving a specific point from a multipoint geometry requires this mess of casts!